Fix #3277: Created BUILD.bazel for domain/state

yash10019coder opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #3277 
Created a BUILD.bazel file in domain state
## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

yash10019coder committed:
Created BUILD.bazel for domain/state

yash10019coder commented:
@fsharpasharp @anandwana001 @prayutsu  created a PR but unable to add dependencies in BUILD.bazel please help

yash10019coder committed:
updated BUILD.bazel

yash10019coder committed:
Added the directory to the MIGRATED_PROD_FILES of the module's root BUILD.bazel file

yash10019coder commented:
All the dependencies in the domain/utility classes are of model and it has no bazel packages for it therefore I'm haveing trouble including the dependencies of the domain/utility classes

BenHenning commented:
@yash10019coder you'll need to add dependencies from https://github.com/oppia/oppia-android/blob/develop/model/BUILD.bazel, e.g.: ``//model:exploration_java_proto_lite`` for each of the protos that are used by these classes. I think Exploration gives you most of them, but you might need a few more.

yash10019coder committed:
Added dependencies in bazel/state file and added dependencies of the bazel/state in global of bazel/domain in kt_android_library

yash10019coder commented:
> @yash10019coder you'll need to add dependencies from https://github.com/oppia/oppia-android/blob/develop/model/BUILD.bazel, e.g.: `//model:exploration_java_proto_lite` for each of the protos that are used by these classes. I think Exploration gives you most of them, but you might need a few more.

@BenHenning I have added these dependencies in my latest commit but I still get some errors and how did you found out that we have to put this in dependencies //model:exploration_java_proto_lite three is no mention of the java files in it

and I'm attaching the screenshot of the error

![Screenshot from 2021-08-21 10-13-22](https://user-images.githubusercontent.com/76404817/130310953-9873d398-37fa-414c-9564-641f9c067850.png)

prayutsu commented:
> > @yash10019coder you'll need to add dependencies from https://github.com/oppia/oppia-android/blob/develop/model/BUILD.bazel, e.g.: `//model:exploration_java_proto_lite` for each of the protos that are used by these classes. I think Exploration gives you most of them, but you might need a few more.
> 
> @BenHenning I have added these dependencies in my latest commit but I still get some errors and how did you found out that we have to put this in dependencies //model:exploration_java_proto_lite three is no mention of the java files in it
> 
> and I'm attaching the screenshot of the error
> 
> ![Screenshot from 2021-08-21 10-13-22](https://user-images.githubusercontent.com/76404817/130310953-9873d398-37fa-414c-9564-641f9c067850.png)

@yash10019coder To find out what deps need to be added, you can check the imports of the file `StateDeck.kt` as imports tell us what other classes are we using in that file.
If you look at the error carefully, it says `unresolved reference` which means that Bazel does not know about those particular classes when it tries to build these newly introduces libraries. If you want to tell Bazel about these classes, then you need to add those classes in the `deps` as well.

prayutsu commented on .idea/runConfigurations.xml:
```diff
@@ -1,12 +0,0 @@
-      </set>
-    </option>
-  </component>
-</project>
```
Please restore this file.

BenHenning replied:
It appears this file was removed again--please restore it.

prayutsu commented on domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel:
```diff
@@ -0,0 +1,38 @@
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//model:exploration_java_proto_lite"
+    ]
```
Add all missing deps that are being highlighted in the build error. 
Ditto for other added libraries.

yash10019coder replied:
@prayutsu  that I have understood but the main problem is how to find which proto is to be added
like AnswersAndResponse.java is being used in StateDeck but now how can I find the particular of this in bazel/model

prayutsu replied:
You need to find the proto file that contains the message `AnswerAndResponse` in it and then you can specify the library of that file here.
To find the .proto file just press `Ctrl + Shift + R` and type the keywords that you want to search in the whole codebase.

prayutsu reviewed (changes requested):
@yash10019coder Left some comments, PTAL.

yash10019coder committed:
Added runConfigurations.xml file

prayutsu reviewed (commented):
@yash10019coder Followed up on our comment.

yash10019coder committed:
Added all the dependencies in bazel/state and added api visiblity to some java_proto_lite in bazel/model

yash10019coder commented:
@prayutsu @fsharpasharp made a new commit PTAL 
I'm still getting errors althought I have added all the dependencies 
attaching the SS below

![Screenshot from 2021-08-22 10-10-23](https://user-images.githubusercontent.com/76404817/130342643-6abd6215-3dd6-461b-98f2-8d8086d5199d.png)
![Screenshot from 2021-08-22 10-10-37](https://user-images.githubusercontent.com/76404817/130342646-e281bd04-cf38-424a-9a87-25e4baf9caa9.png)
![Screenshot from 2021-08-22 10-10-51](https://user-images.githubusercontent.com/76404817/130342647-ad0c9ca9-dc11-4bb8-bc38-4396b58f95e7.png)
![Screenshot from 2021-08-22 10-11-12](https://user-images.githubusercontent.com/76404817/130342648-3c36f332-e3f8-47bd-b520-ebd9db7ebb67.png)
![Screenshot from 2021-08-22 10-11-21](https://user-images.githubusercontent.com/76404817/130342649-cc6f1cb5-7daf-441c-b05c-eec025c49db9.png)
![Screenshot from 2021-08-22 10-11-28](https://user-images.githubusercontent.com/76404817/130342650-d80ed567-e9fb-4b5a-9d03-614323ac65fe.png)

yash10019coder committed:
reverted bazel/model

yash10019coder committed:
Fixed all the errors now bazel/state file successfully builds
Also removed the internal keyword from StateDeck.kt StateGraph.kt and StateList.kt

yash10019coder commented:
@prayutsu @fsharpasharp @BenHenning PTAL

yash10019coder committed:
Merge branch 'develop' into create-build-bazel-domain-state

yash10019coder committed:
Added a one more dependency to the bazel/state list and graph in bazel/state

yash10019coder commented:
Made a commit @prayutsu @fsharpasharp @BenHenning 
Still getting the error PTAL and here's the SS
![Screenshot from 2021-08-23 12-11-58](https://user-images.githubusercontent.com/76404817/130402614-ac176e62-8c4e-4d7f-9d02-95afe511d7f2.png)
![Screenshot from 2021-08-23 12-14-46](https://user-images.githubusercontent.com/76404817/130402622-d211d157-1371-4647-b345-e990b478a94e.png)

BenHenning commented on domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel:
```diff
@@ -0,0 +1,57 @@
+    ],
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//model:exploration_checkpoint_java_proto_lite",
+        "//model:exploration_java_proto_lite",
+        "//model:feedback_reporting_java_proto_lite",
+        "//model:interaction_object_java_proto_lite",
+        "//model:onboarding_java_proto_lite",
+        "//model:question_java_proto_lite",
+        "//model:subtitled_html_java_proto_lite",
+        "//model:topic_java_proto_lite",
```
We shouldn't be adding all of these--you should look at the import statements within StateDeck to see which model dependencies are needed, then add them here & build locally to verify.

Ditto for the other new libraries below.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel:
```diff
@@ -0,0 +1,57 @@
+"""
+build bazel for domain state
```
Suggest instead:

```
Domain services & definitions corresponding to managing ephemeral states in play sessions.
```

BenHenning commented on domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt:
```diff
@@ -85,7 +85,7 @@ internal class StateDeck internal constructor(
   }
 
   /** Returns the current [EphemeralState] the learner is viewing. */
-  internal fun getCurrentEphemeralState(helpIndex: HelpIndex): EphemeralState {
```
Did you mean to remove this parameter? It seems like it's still needed.

yash10019coder replied:
Actually I didn't mean to remove the param I wanted to remove the internal keyword it was removed by mistake

BenHenning commented on domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt:
```diff
@@ -128,14 +128,53 @@ internal class StateDeck internal constructor(
+    solutionIsRevealed = false
+  }
+
+  fun pushStateForHint(state: State, hintIndex: Int): EphemeralState {
+    val interactionBuilder = state.interaction.toBuilder().setHint(
+      hintIndex,
+      hintList.get(0)
+    )
+    val newState = state.toBuilder().setInteraction(interactionBuilder).build()
+    val ephemeralState = EphemeralState.newBuilder()
+      .setState(newState)
+      .setHasPreviousState(!isCurrentStateInitial())
+      .setPendingState(
+        PendingState.newBuilder().addAllWrongAnswer(currentDialogInteractions).addAllHint(hintList)
+      )
+      .build()
+    pendingTopState = newState
+    hintList.clear()
+    // Increment the value of revealHintIndex by 1 every-time a new hint is revealed.
+    revealedHintIndex++
+    return ephemeralState
+  }
+
+  fun pushStateForSolution(state: State): EphemeralState {
+    val interactionBuilder = state.interaction.toBuilder().setSolution(solution)
+    val newState = state.toBuilder().setInteraction(interactionBuilder).build()
+    val ephemeralState = EphemeralState.newBuilder()
+      .setState(newState)
+      .setHasPreviousState(!isCurrentStateInitial())
+      .setPendingState(
+        PendingState.newBuilder().addAllWrongAnswer(currentDialogInteractions).addAllHint(hintList)
+      )
+      .build()
+    pendingTopState = newState
+    solutionIsRevealed = true
+    return ephemeralState
```
Did these come from a merge conflict? I don't think we want to reintroduce this code.

In general, please make sure to double check your changes before pushing to ensure exactly the changes you intend to submit are pushed.

Ditto elsewhere in this & the other changed files.

yash10019coder replied:
Yes there was a merge confilct

yash10019coder replied:
Shall I remove this or let it be ?

BenHenning replied:
I think you've fixed it in the latest change. All I see are the removals of 'internal' which seems correct.

BenHenning commented on model/BUILD.bazel:
```diff
@@ -178,6 +179,7 @@ format_import_proto_library(
 
 java_lite_proto_library(
     name = "feedback_reporting_java_proto_lite",
+    visibility = ["//:oppia_api_visibility"],
```
Here & above: I suspect these visibility changes aren't actually needed.

BenHenning reviewed (changes requested):
Thanks @yash10019coder. I think a bunch of extra changes got pulled in, maybe from a recent merge?

PTAL at latest comments.

BenHenning commented:
@yash10019coder the deps you added should result in no build failures (though there is extra code that's been added). Can you try addressing all the comments in this PR & then build again to see if the problem gets addressed?

yash10019coder committed:
Added runConfiguration.xml file

yash10019coder committed:
Restored the files to their orignal code StateDeck.kt StateGraph.kt StateList.kt

yash10019coder committed:
removed extra dependencies from bazel/state

yash10019coder committed:
removed internal keyword form all the kotlin files

yash10019coder committed:
removed api visiblity form model/bazel

yash10019coder committed:
Added api visibility to the question_java_proto_lite in model/bazel

yash10019coder commented:
@prayutsu @BenHenning made a PR PTAL

BenHenning commented:
Hi @yash10019coder. My apologies for the delay--I'll need to take a look at this tomorrow.

BenHenning commented:
Hi--sorry again about the delay. I still plan to look at this later in the week; I've just had some project emergencies to deal with in the meantime. Thanks for your patience.

yash10019coder commented:
@BenHenning please review the PR it is marked staled

BenHenning commented on .idea/runConfigurations.xml:
```diff
@@ -3,9 +3,12 @@
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.AllInPackageGradleConfigurationProducer" />
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.TestClassGradleConfigurationProducer" />
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.TestMethodGradleConfigurationProducer" />
+        <option
+          value="org.jetbrains.plugins.gradle.execution.test.runner.AllInPackageGradleConfigurationProducer" />
+        <option
+          value="org.jetbrains.plugins.gradle.execution.test.runner.TestClassGradleConfigurationProducer" />
+        <option
+          value="org.jetbrains.plugins.gradle.execution.test.runner.TestMethodGradleConfigurationProducer" />
```
Please revert these changes--they're unrelated to the PR.

BenHenning reviewed (commented):
Thanks @yash10019coder. So sorry for the extended delay--I will prioritize reviewing this this week.

I had one new comment, and I think one of my earlier comments hasn't yet been resolved. Please make sure to address & reply to all comments before sending a PR back to be re-reviewed.

yash10019coder committed:
Reverted runConfigurations.xml

yash10019coder committed:
Update BUILD.bazel

yash10019coder reviewed (commented):
@BenHenning resolved the comments

yash10019coder commented:
@BenHenning PTAL

BenHenning commented on domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel:
```diff
@@ -0,0 +1,44 @@
+    ],
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//model:event_logger_java_proto_lite",
+        "//model:exploration_checkpoint_java_proto_lite",
```
Are these two needed?

yash10019coder replied:
Yes I think it is needed

yash10019coder replied:
import org.oppia.android.app.model.Outcome
because of this import on line 4 in StateList.kt

yash10019coder replied:
import org.oppia.android.app.model.State
on line 6 in StateList.kt

BenHenning replied:
We should be using exploration_java_proto_lite, then, for those (not the checkpoint version of it).

BenHenning reviewed (commented):
Thanks @yash10019coder. Just 2 new comments.

Similar to my comment in your other PR, please make sure to respond via a comment to all conversation threads from reviewers to confirm you saw them & addressed/didn't address them.

yash10019coder committed:
Merge branch 'develop' into create-build-bazel-domain-state

BenHenning reviewed (commented):
Thanks @yash10019coder. Justhad one follow-up.

yash10019coder committed:
removed unused and unnecessary dependencies

yash10019coder committed:
Merge branch 'create-build-bazel-domain-state' of github.com:yash10019coder/oppia-android into create-build-bazel-domain-state

yash10019coder commented:
@BenHenning removed unused and unncesscarty imports from the bazel build file PTAL

yash10019coder committed:
Update BUILD.bazel

Added some deps in glob due to failing tests these changes happened due to merge conflict

BenHenning reviewed (approved):
Thanks @yash10019coder. This LGTM!

BenHenning commented:
@prayutsu it seems you requested changes. Could you PTAL?

prayutsu reviewed (approved):
LGTM, Thanks @yash10019coder

prayutsu commented:
@BenHenning This is ready to merge now 😄

BenHenning commented:
Thanks @prayutsu & @yash10019coder.
